Added Extension Sorting capability according to downloads and last published date

saurabh95 opened the pull request:
Now on opening the Extension Manager, a drop-down is shown on the left side of the search bar, which indicates the sorting of extensions according to the selected criteria(currently sort methods available are  Last Updated and Download Count), on change of the selection we store the the changed selection in the preferences and if the Extension Manager is invoked again, then we use the same sorting criteria which was selected by user earlier. The sorting is available for Available and Theme Extensions.
I have attached a snapshot of the feature in use where the sort order and the download count has been highlighted for illustration. 

![downloadcount](https://cloud.githubusercontent.com/assets/5630113/22639343/937ad5d8-ec73-11e6-89aa-99ba11475ab1.png)

saurabh95 committed:
Added Extension Sorting capability according to downloads and last published date

petetnt commented on src/extensibility/ExtensionManagerViewModel.js:
```diff
@@ -292,6 +292,20 @@ define(function (require, exports, module) {
 
+    ExtensionManagerViewModel.prototype._setSortedExtensionList = function (extensions, isTheme) {
+        this.filterSet = this.sortedFullSet = registry_utils.sortRegistry(extensions, "registryInfo")
+            .filter(function (entry) {
```
There seems to be some redundancy here, seems like same could be achieved with:

``` js
return entry.registryInfo && entry.registryInfo.metadata;
```

Do correct me if I am wrong 👍

saurabh95 replied:
yeah that can be changed :)

petetnt reviewed (commented):
Mostly LGTM, needs tests 👍

ficristo commented:
Nice! I think @zaggino and @dnbard could have some opinions on this.
See @dnbard extension: https://github.com/dnbard/brackets-extension-rating

I have two doubts:
- why we should threat the installed tab differently?
- `registry_utils` comes from brackets-registry: I guess it is necessary to change it but it seems to me it should be more "self contained" (= don't `require` Brackets code)

saurabh95 commented:
@ficristo the installed tab is treated differently as we do not have the download count of the installed extensions in the installInfo, and also currently the sorting of installed extensions is done according to the update available.
Yeah we can isolate the registry_utils from Brackets code, we can pass the preference as the argument.

saurabh95 committed:
Passed sort criteria to sortRegistry function in registy_utils

saurabh95 committed:
Added some tests

saurabh95 commented:
@ficristo @petetnt Addressed review comments and also added some tests.

saurabh95 committed:
Added check for existence of downloadCount in extension-manager

petetnt commented:
Great stuff @saurabh95!

Some issues I spotted:

1. For me the extension manager opens on Installed page and the sort by-toggle is visible (and does nothing). Needs similar check to https://github.com/adobe/brackets/pull/13080/files#diff-3491ce7bb262eb073c8aabc3d3e064c0R376 when opening the window too.

2. The sort toggle breaks the layout with really narrow windows (but the search bar does it too). Feel free to disregard / bump to another issue.
![image](https://cloud.githubusercontent.com/assets/7641760/22806174/757b6da0-ef29-11e6-954d-5fb4b518955b.png)

3. The icons for download counts don't seem to work on Windows (10)
![image](https://cloud.githubusercontent.com/assets/7641760/22806196/8bd960b6-ef29-11e6-9f85-dbb7de57ceb2.png)

petetnt commented on test/spec/ExtensionManager-test.js:
```diff
@@ -829,6 +838,12 @@ define(function (require, exports, module) {
+                    PreferencesManager.set("extensions.sort", "downloadCount");
+                    model._setSortedExtensionList(ExtensionManager.extensions, true);
+                    expect(model.filterSet).toEqual(["theme-2", "theme-1"]);
+                });
```
A simple check could be added that Brackets actually renders the `totalDownloads` to the UI 👍

petetnt reviewed (changes requested):
Reviewed again, couple of small issues, otherwise LGTM 👍

ficristo commented on src/extensibility/registry_utils.js:
```diff
@@ -136,8 +136,16 @@ define(function (require, exports, module) {
         sortedEntries.sort(function (entry1, entry2) {
-            return getPublishTime((subkey && entry2[subkey]) || entry2) -
-                getPublishTime((subkey && entry1[subkey]) || entry1);
+            if (sortBy !== "publishedDate") {
```
Not a big deal, but in general I prefer if you spell out all cases:
```js
if (sortBy === "downloadCount") {
} else if (sortBy === "publishedDate") {
} else {
  console.error("Unknown sortBy " + sortBy ". Defaults to....")
}
```
Or something like this.

ficristo reviewed (approved):
I didn't test this so I trust @petetnt.
But in general seems good.

saurabh95 committed:
Added test for checking downloadCount in ExtensionManager view

saurabh95 commented:
@petetnt I have added a test for checking downloadCount in view and also added a check for active tab on launch of Extension Manager to hide/show the sort option, regarding the download icon, I will replace it with an image.

saurabh95 committed:
Added svg image for download icon in Extension Manager

swmitra reviewed (approved):
Good stuff @saurabh95 👍

saurabh95 commented:
A screenshot with the new icon. 

![untitled](https://cloud.githubusercontent.com/assets/5630113/23113927/9bf40254-f761-11e6-9ce8-83ddfb7c5ba4.png)

swmitra commented:
Merging Now with the new Icon.
